src/main/webapp/app/components/tabs/curation-data-import-tab/import-methods.tsx

```diff
@@ -0,0 +1,286 @@
+      if (propValue !== valData.val()) {
+        await firebaseGeneReviewService.updateReviewableContent(valPath, valData.val(), propValue, reviewData.val(), uuidData.val());
+        status.status = 'complete';
+      } else {
+        status.status = 'warning';
+        status.message = 'The data is identical to production, no update needed.';
+      }
+    } else {
+      status.status = 'error';
+      status.message = `The gene property \`${genePropKey}\`does not exist. Skip importing`;
+    }
+    return Promise.resolve(status);
+  });
+};
+
+export const saveGenomicIndicator = async (
+  firebaseGeneService: FirebaseGeneService,
+  dataRow: DataRow<GenomicIndicatorDI>,
+  hugoSymbol: string,
+  isGermline: boolean,
+) => {
+  return geneCheck(firebaseGeneService, isGermline, hugoSymbol, async () => {
+    const genomicIndicatorName = dataRow.data.genomic_indicator.trim();
+    const genePath = getFirebaseGenePath(isGermline, hugoSymbol);
+    const gipath = genePath + '/genomic_indicators';
+    const giData = await firebaseGeneService.getObject(gipath);
+    const alleleStates = (dataRow.data.allele_state || '')
+      .split(',')
+      .map(state => state.trim())
+      .filter(state => !!state);
+    return Promise.resolve(
+      await alleleStateCheck(alleleStates, async () => {
+        let status = new DataImportStatus();
+        if (giData.exists()) {
+          const genomicIndicators: GenomicIndicator[] = giData.val();
+          if (genomicIndicators.filter(gi => gi.name.toLowerCase() === genomicIndicatorName.toLowerCase()).length > 0) {
+            status.status = 'error';
+            status.message = 'Genomic indicator already exists';
+          } else {
+            status = await saveGenomicIndicatorToFirebase(
+              firebaseGeneService,
+              gipath,
+              genomicIndicatorName,
+              dataRow.data.description,
+              alleleStates as ALLELE_STATE[],
+            );
+          }
+        } else {
+          status = await saveGenomicIndicatorToFirebase(
+            firebaseGeneService,
+            gipath,
+            genomicIndicatorName,
+            dataRow.data.description,
+            alleleStates as ALLELE_STATE[],
+          );
+        }
+        return Promise.resolve(status);
+      }),
+    );
+  });
+};
+
+const saveGenomicIndicatorToFirebase = async (
+  firebaseGeneService: FirebaseGeneService,
+  genomicIndicatorsPath: string,
+  name: string,
+  description?: string,
+  alleleStates?: ALLELE_STATE[],
+): Promise<DataImportStatus> => {
+  await firebaseGeneService.addGenomicIndicator(true, genomicIndicatorsPath, name, description, alleleStates);
+  return Promise.resolve({
+    status: 'complete',
+    message: '',
+  });
+};
+export const saveMutation = async (
+  authStore: AuthStore,
+  firebaseGeneService: FirebaseGeneService,
+  firebaseMetaService: FirebaseMetaService,
+  isGermline: boolean,
+  dataRow: DataRow<GermlineMutationDI | SomaticMutationDI>,
```

Comment:
Because this method is used for both germline and somatic, the row format has two. But because of this, I have to force the type conversion in line 220 and 231. Any suggestions? @jfkonecn

Reply:
Does this work?

```ts
  const isGermlineData = (data: GermlineMutationDI | SomaticMutationDI): data is GermlineMutationDI => {
    return isGermline;
  };

  if (isGermlineData(dataRow.data)) {
    const data = dataRow.data;
    if (data.pathogenicity) {
      const importStatus: DataImportStatus = pathogenicityCheck(data.pathogenicity);
      if (importStatus.status === 'error') {
        return Promise.resolve(importStatus);
      }
    }
    mutation.mutation_effect.pathogenic = data.pathogenicity as PATHOGENICITY;
    mutation.mutation_effect.pathogenic_review = new Review(authStore.fullName, '');
    mutationImpactStatusUpdated = true;
  } else {
    const data: SomaticMutationDI = dataRow.data;
    if (data.oncogenicity) {
      const importStatus: DataImportStatus = oncogenicityCheck(data.oncogenicity);
      if (importStatus.status === 'error') {
        return Promise.resolve(importStatus);
      }
    }
    mutation.mutation_effect.oncogenic = data.oncogenicity as FIREBASE_ONCOGENICITY;
    mutation.mutation_effect.oncogenic_review = new Review(authStore.fullName, '');
    mutationImpactStatusUpdated = true;
  }
```

Reply:
Yeah I agree. I tried other techniques, but this was the only one that seemed to make typescript happy.